File: Stacks/Patterns/Pattern7/Pattern7.py
def nse(arr):
    n = len(arr)
    nse_arr = [0] * n
    stk = []
    for i in range(n - 1, -1, -1):
        while stk and arr[stk[-1]] > arr[i]:
            stk.pop()
        if not stk:
            nse_arr[i] = n
        else:
            nse_arr[i] = stk[-1]
        stk.append(i)
    return nse_arr


def pse(arr):
    n = len(arr)
    pse_arr = [0] * n
    stk = []
    for i in range(n):
        while stk and arr[stk[-1]] >= arr[i]:
            stk.pop()
        if not stk:
            pse_arr[i] = -1
        else:
            pse_arr[i] = stk[-1]
        stk.append(i)
    return pse_arr


nums = [2, 2, 2]
nse_arr = nse(nums)
pse_arr = pse(nums)
sum_val = 0
for i in range(len(nums)):
    left = i - pse_arr[i]
    right = nse_arr[i] - i
    sum_val += left * right * nums[i]
print(sum_val)

File: Stacks/Patterns/Pattern7/Pattern7.java
import java.util.*;
public class Pattern7 {
    public static void main(String[] args) {
        int[] nums={2,2,2};
        int[] nsearr=new int[nums.length];
        int[] psearr=new int[nums.length];
        
        nsearr=nse(nums,nsearr,new Stack<>());
        psearr=pse(nums,psearr,new Stack<>());
        System.out.println(Arrays.toString(nsearr));
        System.out.println(Arrays.toString(psearr));
        int sum=0;
        for(int i=0;i<nums.length;i++){
            int left=i-psearr[i];
            int right=nsearr[i]-i;

            sum+=(left*right*nums[i]);
        }
        System.out.println(sum);
    }

    public static int[] nse(int[] arr, int[] nse_arr, Stack<Integer> stk){
        for(int i=arr.length-1;i>=0;i--){
            while(!stk.empty() && arr[stk.peek()]>arr[i]){
                stk.pop();
            }
            if(stk.empty()){
                nse_arr[i]=arr.length;
            }
            else{
                nse_arr[i]=stk.peek();
            }
            stk.push(i);
        }
        return nse_arr;
    }
    public static int[] pse(int[] num,int[] pse_num,Stack<Integer> stk){

        for (int i=0;i<num.length;i++){
            while(!stk.empty() && num[stk.peek()]>=num[i]){
                stk.pop();
            }
            if(stk.empty()){
                pse_num[i]=-1;
            }
            else{
                pse_num[i]=stk.peek();
            }
            stk.push(i);
        }
        return pse_num;
    }
}
